ios/LayoutReanimation/REASharedTransitionManager.m

```diff
@@ -0,0 +1,518 @@
+#import <RNReanimated/REASharedElement.h>
+#import <RNReanimated/REASharedTransitionManager.h>
+#import <objc/runtime.h>
+
+#define LOAD_SCREENS_HEADERS ((!RCT_NEW_ARCH_ENABLED && __has_include(<RNScreens/RNSScreen.h>)) || (RCT_NEW_ARCH_ENABLED && __has_include(<RNScreens/RNSScreen.h>) && __cplusplus))
+
+#if LOAD_SCREENS_HEADERS
+#import <RNScreens/RNSScreen.h>
+#import <RNScreens/RNSScreenStack.h>
+#endif
+
+@implementation REASharedTransitionManager {
+  NSMutableDictionary<NSNumber *, UIView *> *_sharedTransitionParent;
+  NSMutableDictionary<NSNumber *, NSNumber *> *_sharedTransitionInParentIndex;
+  NSMutableDictionary<NSNumber *, REASnapshot *> *_snapshotRegistry;
+  NSMutableDictionary<NSNumber *, UIView *> *_currentSharedTransitionViews;
+  REAFindPrecedingViewTagForTransitionBlock _findPrecedingViewTagForTransition;
+  UIView *_transitionContainer;
+  NSMutableArray<UIView *> *_addedSharedViews;
+  BOOL _isSharedTransitionActive;
+  NSMutableArray<REASharedElement *> *_sharedElements;
+  REAAnimationsManager *_animationManager;
+  NSMutableSet<NSNumber *> *_screenHasObserver;
+}
+
+- (instancetype)initWithAnimationsManager:(REAAnimationsManager *)animationManager
+{
+  if (self = [super init]) {
+    _snapshotRegistry = [NSMutableDictionary new];
+    _currentSharedTransitionViews = [NSMutableDictionary new];
+    _addedSharedViews = [NSMutableArray new];
+    _sharedTransitionParent = [NSMutableDictionary new];
+    _sharedTransitionInParentIndex = [NSMutableDictionary new];
+    _isSharedTransitionActive = NO;
+    _sharedElements = [NSMutableArray new];
+    _animationManager = animationManager;
+    _screenHasObserver = [NSMutableSet new];
+  }
+  return self;
+}
+
+- (void)invalidate
+{
+  _snapshotRegistry = nil;
+  _currentSharedTransitionViews = nil;
+  _addedSharedViews = nil;
+  _sharedTransitionParent = nil;
+  _sharedTransitionInParentIndex = nil;
+  _sharedElements = nil;
+  _animationManager = nil;
+}
+
+- (UIView *)getTransitioningView:(NSNumber *)tag
+{
+  return _currentSharedTransitionViews[tag];
+}
+
+- (void)notifyAboutNewView:(UIView *)view
+{
+  [_addedSharedViews addObject:view];
+}
+
+- (void)viewsDidLayout
+{
+  [self setupAsyncSharedTransitionForViews:_addedSharedViews];
+  [_addedSharedViews removeAllObjects];
+}
+
+- (void)setupAsyncSharedTransitionForViews:(NSArray<UIView *> *)views
+{
+  NSArray *sharedViews = [self sortViewsByTags:views];
+  _sharedElements = [self getSharedElementForCurrentTransition:sharedViews withNewElements:YES];
+}
+
+- (BOOL)setupSyncSharedTransitionForViews:(NSArray<UIView *> *)views
```

Comment:
why is "sync" part of the name?

Reply:
I changed this name because was misleading. Resolved by: https://github.com/software-mansion/react-native-reanimated/pull/4014/commits/06ff48a085c6d2d8fe3704857b2b283444598877